### DISCUSSION TITLE
Редактировать текст комментариев написанного пользователем

### DISCUSSION THREAD

**Discussion by vldekov** [totumonline/totum-ru-issues-and-discussions#369] [Как это сделать?] [answered]
Есть возможность программно редактировать комментарии?

**Comment by erpic-ru** [comment]
Да: https://ru.docs.totum.online/fields#type-comments

**Reply by vldekov** [reply]
Перезаписать полностью лист комментариев:
example2: set(table: $#ntn; field: 'h_kommentariy' = $list)
list: listCreate(item: $listComment1; item: $listComment2)
listComment1: listCreate(item: "2019-12-12 12:12"; item: 1; item: "Комментарий 1")
listComment2: listCreate(item: "2019-12-12 12:13"; item: 1; item: "Комментарий 2")

Требуется редактировать только последний комментарий.

**Reply by erpic-ru** [reply]
Нужно посчитать количество комментариев в списке  через listcount, взять из него последнюю запись, поменять и, засетить в него то что было с измененным последним комментарием. Либо удалить последнюю и добавить новую.
Вот код по доставанию значения последнего комментария. Остальное можно додумать
```
1:listCount(list: #comment)
por: listItem(list: #comment; item: $1 - 1)
comment: $por[2]
```

**Comment by vldekov** [comment]
ch1:set(table: $#ntn; field: 'test' = str`$sel3++"( тест )"`; where: 'nomerpaz' = #nomerpaz; log: true)
sel2:listCount(list: #test)-1
sel3:str`#test[$sel2][2]`
Не работает. Т.к. требуется редактировать только последний комментарий. Как надо написать SET?

**Reply by erpic-ru** [reply]
Там весь массив c комментариями нужно передать. Попробуйте создать поле Данные получить в него кодом данные из поля с комментариями и посмотреть через панель с Json как там все устроено, затем уже пробовать в нем что-то поменять

**Comment by vldekov** [comment]
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/114401808/ba0c2628-4d5f-4c0f-b3c9-a14d3ddd9dc7)

**Reply by vldekov** [reply]
При использовании этого примера система почему то подвисает...

**Reply by ttm-alexeypolunin** [reply]
В этой ветке разобрались?

**Comment by vldekov** [comment]
Как обратиться к последнему полю "комментарии" в контексте функции SET?

![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/114401808/5b7c82d1-6d3c-4bd8-b761-3a72d4b803ec)

**Reply by erpic-ru** [reply]
Можете попробовать через [listreplace перебрать и заменить](https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/316).

**Comment by vldekov** [comment]
Пока не получается...( Продолжаем пробовать... 05.06.2023, 14:22, "Alexey Polunin" ***@***.***>: В этой ветке разобрались?—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Вы можете переписать все содержимое поля Комментарий. Внутри там простая структура — посмотрите ее через linkToDataJson.

У пользователя через интерфейс есть возможность отредактировать последний комментарий. 

если вам нужно изменить комент в середине, то вам нужно перебрать весь массив и найдя по условиям нужный коммент изменить его содержимое на новое, оставляя остальные комментарии без изменений:

```
= : set(table: $#ntn; field: 'h_comments' = $edited; log: true)

edited: listReplace(list: #h_comments; action: "2" = $val; key: "key"; value: "value")

	val: if(condition: $#key = 1; then: "Some edited comment"; else: $#value[2])
```

**Reply by vldekov** [reply]
Попробовал использовать этот код на изменение. Записывает почему то отредактированный комментарий в другую запись. Уточню задачу. Необходимо программно дописывать в каждый текущий ( новый ) комментарий пользователя фразу "(передано)".

**Reply by erpic-ru** [reply]
Как быстрый и простой вариант - добавьте linktoinput с кодом set в поле с комментариями, с добавлением к значению в input того, что вам нужно. Если прям нужно в комментариях это все делать, то вариант выше нужно допилить с кодом на изменение поля: вычислять последний и править

**Reply by ttm-alexeypolunin** [reply]
я на уровне описания задачи не понимаю, что вам нужно сделать — запишитесь хотя бы на бесплатную консультацию, расскажете. Этой задачи там максимум на 5 мин.

**Reply by vldekov** [reply]
Необходимо в момент сохранения пользователем своего комментария дописывать в его комментарий программно ( кодом действия на изменение ) текст "(передано)". Например, пользователь написал комментарий "Документы разместил.", а программа изменила на текст "Документы разместил. (передано)". Или например, пользователь написал комментарий "Документы разместил.", а программа изменила на текст "Документы разместил. (2023,06,26 09:00)"

**Reply by ttm-alexeypolunin** [reply]
Не знаю зачем вам это нужно конешн:

```
=: set(table: $#ntn; field: $#nf = $edited; log: true)

edited: listReplace(list: $#nfv; action: "2" = $val; key: "key"; value: "value")

	~last: listCount(list: $#nfv) - 1

	val: if(condition: $#key = $last; condition: $reg != true; then: str`$#value[2] ++ "(передано)"`; else: $#value[2])
	
		reg: strRegMatches(str: $#value[2]; template: "(передано)"; matches: "matches"; flags: "u")
```

Код действия при изменении в поле Комментарий

**Reply by vldekov** [reply]
Этот код работает не корректно. Почему то перезаписывает и другие комментарии в других записях таблицы.

**Reply by ttm-alexeypolunin** [reply]
Ну комон - это пример для хэдера. Допишите в set where: 'id' = #id и будет для строчной части

**Reply by vldekov** [reply]
Спасибо! Работает.
